python/lsst/ctrl/bps/wms/panda/panda_service.py

```diff
@@ -151,6 +176,72 @@ def submit(self, workflow):
         _LOG.info("Submitted into iDDs with request id=%i", request_id)
         workflow.run_id = request_id
 
+    @staticmethod
+    def copy_files_for_distribution(tasks, file_distribution_path):
+        """Brings locally generated pickle files into Cloud for further
+        utilization them on the edge nodes.
+
+        Parameters
+        ----------
+        local_pfns: `list` of `tasks`
+            Tasks that input files needs to be placed for
+            distribution
+
+        file_distribution_path: `str`
+            Path on the edge node accessed storage,
+            including access protocol, bucket name to place files
+            :param tasks:
+        """
+
+        local_pfns = {}
+        direct_IO_files = set()
+        for task in tasks:
+            for file in task.files_used_by_task:
+                if not file.delivered:
+                    local_pfns[file.name] = file.submission_url
+                    if file.direct_IO:
+                        direct_IO_files.add(file.name)
+
+        files_to_copy = {}
+
+        # In case there are folders we iterate over its content
+        for local_pfn in local_pfns.values():
+            if os.path.isdir(local_pfn):
+                files_in_folder = ButlerURI.findFileResources([local_pfn])
+                folder_name = os.path.basename(local_pfn)
+                for file in files_in_folder:
+                    file_name = os.path.basename(file.ospath)
+                    files_to_copy[file] = ButlerURI(os.path.join(file_distribution_path,
```

Comment:
You want to retain the hierarchy in source and destination so I think this is all more complicated than it needs to be. I think something like:
```python
files_to_copy[file] = file_distribution_path.join(file.relative_to(local_pfn.parent()))
```

Then you can dispense with all the file_name and folder_name calculations.